tools/old/compactsnoop.py

```diff
@@ -0,0 +1,381 @@
+    valp->high = watermark[WMARK_HIGH];
+    bpf_probe_read(&valp->free, sizeof(valp->free),
+                   &zone->vm_stat[NR_FREE_PAGES]);
+}
+#endif
+
+int trace_compact_zone_entry(struct pt_regs *ctx, struct zone *zone,
+                             struct compact_control *cc)
+{
+#ifdef EXTNEDED_FIELDS
+    struct val_t val = { .fragindex=-1000 };
+#else
+    struct val_t val = { };
+#endif
+    u64 id = bpf_get_current_pid_tgid();
+    PID_FILTER
+    val.sync = cc->sync;
+    start.update(&id, &val);
+    return 0;
+}
+
+int trace_compaction_suitable_entry(struct pt_regs *ctx, struct zone *zone,
+                                    int order)
+{
+    u64 id = bpf_get_current_pid_tgid();
+    struct val_t *valp = start.lookup(&id);
+    if (valp == NULL) {
+        // missed entry
+        return 0;
+    }
+    valp->nid = zone_to_nid_(zone);
+    valp->idx = zone_idx_(zone);
+    valp->order = order;
+
+#ifdef EXTNEDED_FIELDS
+     get_all_wmark_pages(zone, valp);
+#endif
+
+    return 0;
+}
+
+int trace_fragmentation_index_return(struct pt_regs *ctx)
+{
+    int ret = PT_REGS_RC(ctx);
+    u64 id = bpf_get_current_pid_tgid();
+    struct val_t *valp = start.lookup(&id);
+    if (valp == NULL) {
+        // missed entry
+        return 0;
+    }
+#ifdef EXTNEDED_FIELDS
+    valp->fragindex = ret;
+#endif
+    return 0;
+}
+
+int trace_compaction_suitable_return(struct pt_regs *ctx)
+{
+    int ret = PT_REGS_RC(ctx);
+    u64 id = bpf_get_current_pid_tgid();
+    struct val_t *valp = start.lookup(&id);
+    if (valp == NULL) {
+        // missed entry
+        return 0;
+    }
+    if (ret != COMPACT_CONTINUE)
+        start.delete(&id);
+    else
+        valp->ts = bpf_ktime_get_ns();
+    return 0;
+}
+
+int trace_compact_zone_return(struct pt_regs *ctx)
+{
+    int ret = PT_REGS_RC(ctx);
+    struct data_t data = {};
+    u64 ts = bpf_ktime_get_ns();
+    u64 id = bpf_get_current_pid_tgid();
+    struct val_t *valp = start.lookup(&id);
+    if (valp == NULL) {
+        // missed entry or unsuitable
+        return 0;
+    }
+
+    data.delta = ts - valp->ts;
+    data.ts = ts / 1000;
+    data.pid = id >> 32;
+    data.tid = id;
+    bpf_get_current_comm(&data.comm, sizeof(data.comm));
+    data.nid = valp->nid;
+    data.idx = valp->idx;
+    data.order = valp->order;
+    data.sync = valp->sync;
+
+#ifdef EXTNEDED_FIELDS
+    data.fragindex = valp->fragindex;
+    data.min = valp->min;
+    data.low = valp->low;
+    data.high = valp->high;
+    data.free = valp->free;
+#endif
+
+    data.status = ret;
+    data.stack_id = stack_traces.get_stackid(ctx, BPF_F_REUSE_STACKID);
+
+    events.perf_submit(ctx, &data, sizeof(data));
+
+    start.delete(&id);
+    return 0;
+}
+"""
+
+if args.extended_fields:
+    bpf_text = EXTENDED + bpf_text
+else:
+    bpf_text = NO_EXTENDED + bpf_text
+
+if args.pid:
+    bpf_text = bpf_text.replace(
+        "PID_FILTER", "if (pid != %s) { return 0; }" % args.pid)
```

Comment:
This may not work. I only see `id` is defined in the program.

Reply:
Sorry for this, I'll add it.